test/pylib/scylla_cluster.py

```diff
@@ -0,0 +1,792 @@
+            self.cmd = None
+
+    async def uninstall(self) -> None:
+        """Clear all files left from a stopped server, including the
+        data files and log files."""
+
+        if not self.hostname:
+            return
+        logging.info("Uninstalling server at %s", self.workdir)
+
+        shutil.rmtree(self.workdir)
+        self.log_filename.unlink(missing_ok=True)
+
+        await self.host_registry.release_host(self.hostname)
+        self.hostname = ""
+
+    def write_log_marker(self, msg) -> None:
+        self.log_file.seek(0, 2)  # seek to file end
+        self.log_file.write(msg.encode())
+        self.log_file.flush()
+
+    def __str__(self):
+        return self.hostname
+
+
+class ScyllaCluster:
+    def __init__(self, scylla_exe: str, replicas: int, test_base_dir: str,
+                 cmdline_options: List[str], config_options: Dict[str, str],
+                 host_registry: HostRegistry, artifact_registry: ArtifactRegistry) -> None:
+        self.name = str(uuid.uuid1())
+        self.scylla_exe = scylla_exe
+        self.replicas = replicas
+
+        self.cluster_dir = tempfile.mkdtemp(prefix="cluster-", dir=test_base_dir)
+        self.cmdline_options = cmdline_options
+        self.config_options = config_options
+        self.host_registry = host_registry
+        self.artifact_registry = artifact_registry
+        self.running: Dict[str, ScyllaServer] = {}
+        self.stopped: Dict[str, ScyllaServer] = {}
+        self.removed: Set[str] = set()
+        self.is_running: bool = False
+        self.is_dirty: bool = False
+        self.start_exception: Optional[Exception] = None
+        self.keyspace_count = 0
+        self.last_seed: Optional[str] = None     # id as IP Address like '127.1.2.3'
+
+        async def stop_cluster() -> None:
+            await self.stop()
+
+        async def uninstall_cluster() -> None:
+            await self.uninstall()
+
+        self.stop_artifact = stop_cluster
+        self.uninstall_artifact = uninstall_cluster
+
+    async def install_and_start(self) -> None:
+        try:
+            for i in range(self.replicas):
+                await self.add_server()
+            self.keyspace_count = self._get_keyspace_count()
+        except (RuntimeError, NoHostAvailable, InvalidRequest, OperationTimedOut) as e:
+            # If start fails, swallow the error to throw later,
+            # at test time.
+            self.start_exception = e
+        logging.info("Created cluster %s", self)
+        self.is_dirty = False
+
+    async def uninstall(self) -> None:
+        """Stop running servers, uninstall all servers, and remove API socket"""
+        self.is_dirty = True
+        logging.info("Uninstalling cluster")
+        await self.stop()
+        await asyncio.gather(*(server.uninstall() for server in self.stopped.values()))
+        # Note: self.cluster_dir is removed by suite when all tests pass
```

Comment:
Shouldn't you delete the cluster directory now that you have it?

Reply:
No more cluster directory